### PR TITLE
Add Maintenance Configuration module

### PR DESCRIPTION
**Pull request by kevindelmont**
## PR Checklist

---

- [x] I have added example(s) inside the [./examples/] folder
- [x] I have added the example(s) to the integration test list for [normal (~30 minutes)](./workflows/standalone-scenarios.json) or [long runner >30 minutes](./workflows/standalone-scenarios-longrunners.json)
- [x] I have checked the [coding conventions as per the wiki](https://github.com/aztfmod/terraform-azurerm-caf/wiki)
- [x] I have checked to ensure there aren't other open Pull Requests for the same update/change?

## Description

Add Manages a maintenance configuration module

## Does this introduce a breaking change

- [ ] YES
- [x] NO

### CODE REVIEW

**Comment by kevindelmont** [comment]
@arnaudlh can you please take a look at this PR?

**Comment by arnaudlh** [comment]
> @arnaudlh can you please take a look at this PR?

thanks @kevindelmont - running examples, I'm getting: 
```
╷
│ Error: Unsupported attribute
│ 
│   on ../modules/maintenance/configuration/module.tf line 22, in resource "azurerm_maintenance_configuration" "maintenance_configuration":
│   22:     for_each = var.settings.window != null ? [var.settings.window] : []
│     ├────────────────
│     │ var.settings is object with 4 attributes
│ 
│ This object does not have an attribute named "window".
╵```

File: modules/maintenance/configuration/module.tf [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  resource_group_name      = var.resource_group_name
+  location                 = var.location
+  scope                    = var.scope
+  visibility               = try(var.visibility, null)
+  properties               = try(var.properties, {})  
+  in_guest_user_patch_mode = var.settings.scope == "InGuestPatch" ? var.in_guest_user_patch_mode : try(var.in_guest_user_patch_mode, null)
+
+  
+  dynamic "window" {
+    for_each = var.settings.window != null ? [var.settings.window] : []
```
Comment (arnaudlh) [inline modules/maintenance/configuration/module.tf:22]:
```suggestion
    for_each = try(var.settings.window, null) != null ? [var.settings.window] : []
```

File: examples/maintenance_configuration/200-maintenance-configuration-assignment-vm/configuration.tfvars [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+virtual_machines = {
+
+  # Configuration to deploy a bastion host linux virtual machine
+  example_vm1 = {
+    resource_group_key = "rg1"
+    provision_vm_agent = true
+    # when boot_diagnostics_storage_account_key is empty string "", boot diagnostics will be put on azure managed storage
+    # when boot_diagnostics_storage_account_key is a non-empty string, it needs to point to the key of a user managed storage defined in diagnostic_storage_accounts
+    # if boot_diagnostics_storage_account_key is not defined, but global_settings.resource_defaults.virtual_machines.use_azmanaged_storage_for_boot_diagnostics is true, boot diagnostics will be put on azure managed storage
+    boot_diagnostics_storage_account_key = "bootdiag_region1"
```
Comment (arnaudlh) [inline examples/maintenance_configuration/200-maintenance-configuration-assignment-vm/configuration.tfvars:67]:
```suggestion
```

File: examples/maintenance_configuration/200-maintenance-configuration-assignment-vm/configuration.tfvars [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        size           = "Standard_F2"
+        admin_username = "adminuser"
+
+
+        # Spot VM to save money
+        priority        = "Spot"
+        eviction_policy = "Deallocate"
+
+        patch_mode                         = "AutomaticByPlatform"
+        bypassPlatformChecksOnUserSchedule = true
```
Comment (arnaudlh) [inline examples/maintenance_configuration/200-maintenance-configuration-assignment-vm/configuration.tfvars:99]:
```suggestion
        bypass_platform_safety_checks_on_user_schedule_enabled = true
```

File: modules/compute/virtual_machine/vm_windows.tf [hunk 4]
```diff
@@ -41,31 +41,31 @@ resource "azurerm_windows_virtual_machine" "vm" {
...
-  zone                         = try(each.value.zone, null)
-  secure_boot_enabled          = try(each.value.secure_boot_enabled, null)
-  vtpm_enabled                 = try(each.value.vtpm_enabled, null)
-
+  admin_password                                         = try(each.value.admin_password_key, null) == null ? random_password.admin[local.os_type].result : local.admin_password
+  admin_username                                         = try(each.value.admin_username_key, null) == null ? each.value.admin_username : local.admin_username
+  allow_extension_operations                             = try(each.value.allow_extension_operations, null)
+  availability_set_id                                    = can(each.value.availability_set_key) || can(each.value.availability_set.key) ? var.availability_sets[try(var.client_config.landingzone_key, each.value.availability_set.lz_key)][try(each.value.availability_set_key, each.value.availability_set.key)].id : try(each.value.availability_set.id, each.value.availability_set_id, null)
+  bypass_platform_safety_checks_on_user_schedule_enabled = try(each.value.bypass_platform_safety_checks_on_user_schedule_enabled, false)
```
Comment (arnaudlh) [inline modules/compute/virtual_machine/vm_windows.tf:48]:
```suggestion
  bypass_platform_safety_checks_on_user_schedule_enabled = try(each.value.bypass_platform_safety_checks_on_user_schedule_enabled, null)
```

**Review by arnaudlh — CHANGES_REQUESTED** [review]
thanks for the PR @kevindelmont, please review the suggestion to make it work for windows. can you please add support/example for linux vm? thanks

File: modules/maintenance/configuration/module.tf (hunk 1 above)
Comment (kevindelmont) [reply modules/maintenance/configuration/module.tf]: I added example for linux vm and fixed the maintenance configuration module.
Thanks for your suggestions.

File: examples/maintenance_configuration/200-maintenance-configuration-assignment-vm/configuration.tfvars (hunk 2 above)
Comment (kevindelmont) [reply examples/maintenance_configuration/200-maintenance-configuration-assignment-vm/configuration.tfvars]: Thanks

**Pull request merged**